mava/evaluator.py

```diff
@@ -12,340 +12,179 @@
+        # This loop is important because we don't want too many parallel envs.
+        # So in evaluation we have num_envs parallel envs and loop enough times
+        # so that we do at least `eval_episodes` number of episodes.
+        _, metrics = jax.lax.scan(_episode, key, xs=None, length=episode_loops)
+        metrics: Metrics = jax.tree_map(lambda x: x.reshape(-1), metrics)  # flatten metrics
+        return metrics
 
-    Args:
-    ----
-        eval_env (Environment): The environment used for evaluation.
-        network_apply_fn (Union[ActorApply,RecActorApply]): Creates a policy to sample.
-        config (DictConfig): The configuration settings for the evaluation.
-        use_recurrent_net (bool, optional): Whether to use a rnn. Defaults to False.
-        scanned_rnn (Optional[nn.Module], optional): The rnn module.
-            Required if `use_recurrent_net` is True. Defaults to None.
-
-    Returns:
-    -------
-        Tuple[EvalFn, EvalFn]: A tuple of two evaluation functions:
-        one for use during training and one for absolute metrics.
-
-    Raises:
-    ------
-        AssertionError: If `use_recurrent_net` is True but `scanned_rnn` is not provided.
+    def timed_eval_fn(params: FrozenDict, key: PRNGKey, init_act_state: ActorState) -> Metrics:
+        """Wrapper around eval function to time it and add in steps per second metric."""
+        start_time = time.time()
 
-    """
-    # Check if win rate is required for evaluation.
-    log_win_rate = config.env.log_win_rate
-    # Vmap it over number of agents and create evaluator_fn.
-    if use_recurrent_net:
-        assert scanned_rnn is not None
-        evaluator = get_rnn_evaluator_fn(
-            eval_env,
-            network_apply_fn,  # type: ignore
-            config,
-            scanned_rnn,
-            log_win_rate,
-        )
-        absolute_metric_evaluator = get_rnn_evaluator_fn(
-            eval_env,
-            network_apply_fn,  # type: ignore
-            config,
-            scanned_rnn,
-            log_win_rate,
-            10,
-        )
-    else:
-        evaluator = get_ff_evaluator_fn(
-            eval_env,
-            network_apply_fn,  # type: ignore
-            config,
-            log_win_rate,  # type: ignore
-        )
-        absolute_metric_evaluator = get_ff_evaluator_fn(
-            eval_env,
-            network_apply_fn,  # type: ignore
-            config,
-            log_win_rate,
-            10,  # type: ignore
-        )
+        metrics = jax.pmap(eval_fn)(params, key, init_act_state)
+        metrics: Metrics = jax.block_until_ready(metrics)
+
+        end_time = time.time()
+        total_timesteps = jnp.sum(metrics["episode_length"])
+        metrics["steps_per_second"] = total_timesteps / (end_time - start_time)
+        return metrics
+
+    return timed_eval_fn
+
+
+def make_ff_eval_act_fn(actor_network: nn.Module, config: DictConfig) -> EvalActFn:
```

Comment:
A general note here, I actually think we should make it so that this takes in an apply function instead of the network object. THis way we make things more flexible for the future.

Reply:
Done :+1: